README.md
[![Build status][build-image]][build-url]
[![Tests coverage][cov-image]][cov-url]
[![npm version][npm-image]][npm-url]

# date-from-timezone

## Construct date instances within timezone context or resolve date/time information for given timezone

Having **timezone** and **date & time information** resolve **regular date instance that reflects given time point**, or
having **timezone** and **date instance** resolve **date & time information for given timezone**.

Light implementation which resolves needed data through natively available (in all modern engines, both browsers and Node.js) [Intl.DateTimeFormat](http://www.ecma-international.org/ecma-402/1.0/#sec-12.1).

_If loaded in environment which does not provide `Intl.DateTimeFormat` or of which support is incomplete, module resolves to `null`_

### Examples

#### Construct date instances in specific timezone context

```javascript
var dateFromTimezone = require("date-from-timezone");

var getWarsawDate = dateFromTimezone("Europe/Warsaw");
var getShanghaiDate = dateFromTimezone("Asia/Shanghai");

// Signature of getXDate follows: Date(yearOrTimeValue[, month[, date[, hour[, minutes[, seconds[, milliseconds]]]]])
var warsawNoon = getWarsawDate(2017, 6, 5, 12);
// also e.g.
warsawNoon = getWarsawDate(Date.parse("7/5/2017 12:00:00 PM"));

var shanghaiNoon = getShanghaiDate(2017, 6, 5, 12);

console.log(warsawNoon.toISOString()); // "2017-07-05T10:00:00.000Z" (12PM in Warsaw was at 10AM UTC)
console.log(shanghaiNoon.toISOString()); // "2017-07-05T04:00:00.000Z" (12PM in Shanghai was at 4AM UTC)
```

#### Resolve date & time information for given timezone

```javascript
var getTokenize = require("date-from-timezone/get-tokenize");

var warsawTokenize = getTokenize("Europe/Warsaw");
var shanghaiTokenize = getTokenize("Asia/Shanghai");

console.log(warsawTokenize(new Date(Date.UTC(2017, 6, 5, 10))));
// { year: 2017, month 6, date: 5, hours: 12, minutes: 0, seconds: 0, milliseconds: 0 }

console.log(shanghaiTokenize(new Date(Date.UTC(2017, 6, 5, 4))));
// { year: 2017, month 6, date: 5, hours: 12, minutes: 0, seconds: 0, milliseconds: 0 }

// We can also retrieve date object. It's important to understand that it won't reflect given time
// point (so should not be used for date timestamp comparisons) but can be used as interim value
// to be used to get formatted date string
console.log(String(shanghaiTokenize(new Date(Date.UTC(2017, 6, 5, 4))).dateObject));
// In CEST Timezone it'll log:
// "Wed Jul 05 2017 12:00:00 GMT+0200 (CEST)"
```

### Installation

    $ npm install date-from-timezone

### Tests

    $ npm test

[build-image]: https://github.com/medikoo/date-from-timezone/workflows/Integrate/badge.svg
[build-url]: https://github.com/medikoo/date-from-timezone/actions?query=workflow%3AIntegrate
[cov-image]: https://img.shields.io/codecov/c/github/medikoo/date-from-timezone.svg
[cov-url]: https://codecov.io/gh/medikoo/date-from-timezone
[npm-image]: https://img.shields.io/npm/v/date-from-timezone.svg
[npm-url]: https://www.npmjs.com/package/date-from-timezone


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [3.0.1](https://github.com/medikoo/date-from-timezone/compare/v3.0.0...v3.0.1) (2023-01-17)

### Bug Fixes

- Prevent `RangeError: Maximum call stack size exceeded` error in Node.js v14+. Handle gently `24` hour output from `Intl` ([2b0bbf6](https://github.com/medikoo/date-from-timezone/commit/2b0bbf64dc38fc816837f726ebd6e5ead06a894f))

### Maintenance Improvements

- Document internal logic ([1980ed9](https://github.com/medikoo/date-from-timezone/commit/1980ed92bc62be256da7d8f4a18bd286bff4abed))

# [3.0.0](https://github.com/medikoo/date-from-timezone/compare/v2.2.0...v3.0.0) (2019-02-05)

### Features

- convert to ES2015 ([cf21835](https://github.com/medikoo/date-from-timezone/commit/cf21835))
- support one argument time value input ([eedb5c0](https://github.com/medikoo/date-from-timezone/commit/eedb5c0))
- upgrade to ES2015 ([d7eb57e](https://github.com/medikoo/date-from-timezone/commit/d7eb57e))

### BREAKING CHANGES

- Drop ES5 support

<a name="2.2.0"></a>

# [2.2.0](https://github.com/medikoo/date-from-timezone/compare/v2.1.1...v2.2.0) (2018-01-11)

### Features

- introduce dateObject token ([e8a386a](https://github.com/medikoo/date-from-timezone/commit/e8a386a))

<a name="2.1.1"></a>

## [2.1.1](https://github.com/medikoo/date-from-timezone/compare/v2.1.0...v2.1.1) (2017-08-31)

<a name="2.1.0"></a>

# [2.1.0](https://github.com/medikoo/date-from-timezone/compare/v2.0.2...v2.1.0) (2017-08-29)

### Features

- seclude getTokenize utility ([1865ee2](https://github.com/medikoo/date-from-timezone/commit/1865ee2))

<a name="2.0.2"></a>

## [2.0.2](https://github.com/medikoo/date-from-timezone/compare/v2.0.1...v2.0.2) (2017-06-05)

### Bug Fixes

- recalculation didn't handle milliseconds as expected ([7ce8bd8](https://github.com/medikoo/date-from-timezone/commit/7ce8bd8))

<a name="2.0.1"></a>

## [2.0.1](https://github.com/medikoo/date-from-timezone/compare/v2.0.0...v2.0.1) (2017-06-02)

<a name="2.0.0"></a>

# [2.0.0](https://github.com/medikoo/date-from-timezone/compare/v1.0.1...v2.0.0) (2017-06-01)

### Features

- simplify public interface ([f9b7974](https://github.com/medikoo/date-from-timezone/commit/f9b7974))

### BREAKING CHANGES

- Public API changed.

Before:

var dateFromWarsaw = new DateFromTimezone("Europe/Warsaw");
dateFromWarsaw.getDate(year, month);

After:

var getWarsawDate = dateFromTimezone("Europe/Warsaw");
getWarsawDate(year, month);

<a name="1.0.1"></a>

## [1.0.1](https://github.com/medikoo/date-from-timezone/compare/v1.0.0...v1.0.1) (2017-06-01)

<a name="1.0.0"></a>

# 1.0.0 (2017-06-01)


test/index.js
"use strict";

const test             = require("tape")
    , dateFromTimezone = require("../");

test("Main", t => {
	t.equal(typeof dateFromTimezone, "function", "Supported");
	if (!dateFromTimezone) {
		t.end();
		return;
	}

	t.test("UTC", t => {
		const getDate = dateFromTimezone("UTC");

		t.deepEqual(getDate(2017, 1), new Date(Date.UTC(2017, 1)));
		t.end();
	});

	t.test("Europe/Warsaw", t => {
		const getDate = dateFromTimezone("Europe/Warsaw");

		t.deepEqual(getDate(2017, 1, 2, 18), new Date(Date.UTC(2017, 1, 2, 17)));
		t.deepEqual(getDate(2017, 7, 2, 18), new Date(Date.UTC(2017, 7, 2, 16)), "DST");
		t.end();
	});

	t.test("Asia/Shanghai", t => {
		const getDate = dateFromTimezone("Asia/Shanghai");

		t.deepEqual(getDate(2017, 1, 2, 18), new Date(Date.UTC(2017, 1, 2, 10)));
		t.deepEqual(getDate(new Date(2017, 1, 2, 18)), new Date(Date.UTC(2017, 1, 2, 10)));
		t.end();
	});

	t.test("Milliseconds", t => {
		const getDate = dateFromTimezone("Europe/Warsaw");

		t.deepEqual(
			getDate(2017, 1, 2, 18, 14, 20, 123), new Date(Date.UTC(2017, 1, 2, 17, 14, 20, 123))
		);
		t.end();
	});

	t.test("Invalid usage", t => {
		t.throws(() => { dateFromTimezone("WRONG TIMEZONE"); }, RangeError);

		const getDate = dateFromTimezone("Asia/Shanghai");

		t.throws(() => { getDate(); }, TypeError);
		t.throws(() => { getDate(2014999999999, 99); }, /Invalid Date is not valid Date object/u);
		t.throws(() => { getDate(-10, 1); }, /Invalid arguments: No support for BC years/u);
		t.end();
	});
	t.end();
});


test/get-tokenize.js
"use strict";

const test        = require("tape")
    , getTokenize = require("../get-tokenize");

test("Main", t => {
	t.equal(typeof getTokenize, "function", "Supported");
	if (!getTokenize) {
		t.end();
		return;
	}

	t.test("Memoize", t => {
		t.equal(getTokenize("UTC"), getTokenize("UTC"));
		t.end();
	});

	t.test("UTC", t => {
		const tokenize = getTokenize("UTC");

		t.deepEqual(tokenize(new Date(Date.UTC(2017, 1))), {
			year: 2017,
			month: 1,
			date: 1,
			hours: 0,
			minutes: 0,
			seconds: 0,
			milliseconds: 0
		});
		t.end();
	});

	t.test("Europe/Warsaw", t => {
		const tokenize = getTokenize("Europe/Warsaw");

		t.deepEqual(tokenize(new Date(Date.UTC(2017, 1, 2, 17))), {
			year: 2017,
			month: 1,
			date: 2,
			hours: 18,
			minutes: 0,
			seconds: 0,
			milliseconds: 0
		});
		t.deepEqual(
			tokenize(new Date(Date.UTC(2017, 7, 2, 16))),
			{
				year: 2017,
				month: 7,
				date: 2,
				hours: 18,
				minutes: 0,
				seconds: 0,
				milliseconds: 0
			},
			"DST"
		);
		t.end();
	});

	t.test("Asia/Shanghai", t => {
		const tokenize = getTokenize("Asia/Shanghai");

		t.deepEqual(tokenize(new Date(Date.UTC(2017, 1, 2, 10))), {
			year: 2017,
			month: 1,
			date: 2,
			hours: 18,
			minutes: 0,
			seconds: 0,
			milliseconds: 0
		});
		t.end();
	});

	t.test("Milliseconds", t => {
		const tokenize = getTokenize("Europe/Warsaw");

		t.deepEqual(tokenize(new Date(Date.UTC(2017, 1, 2, 17, 14, 20, 123))), {
			year: 2017,
			month: 1,
			date: 2,
			hours: 18,
			minutes: 14,
			seconds: 20,
			milliseconds: 123
		});
		t.end();
	});

	t.test("Date object", t => {
		const tokenize = getTokenize("Europe/Warsaw")
		    , dateObject = tokenize(new Date(Date.UTC(2017, 1, 2, 17, 14, 20, 123)))
				.dateObject;

		t.ok(dateObject instanceof Date);
		t.deepEqual(
			{
				year: dateObject.getFullYear(),
				month: dateObject.getMonth(),
				date: dateObject.getDate(),
				hours: dateObject.getHours(),
				minutes: dateObject.getMinutes(),
				seconds: dateObject.getSeconds(),
				milliseconds: dateObject.getMilliseconds()
			},
			{
				year: 2017,
				month: 1,
				date: 2,
				hours: 18,
				minutes: 14,
				seconds: 20,
				milliseconds: 123
			}
		);
		t.end();
	});

	t.test("Invalid usage", t => {
		t.throws(() => { getTokenize("WRONG TIMEZONE"); }, RangeError);

		const tokenize = getTokenize("Asia/Shanghai");

		t.throws(() => {
			tokenize(new Date(2014999999999, 99));
		}, /Invalid Date is not valid Date object/);
		t.throws(() => {
			tokenize(new Date(-10, 1));
		}, /Invalid arguments: No support for BC years/);
		t.end();
	});
	t.end();
});


test/get-tokenize/index.js
"use strict";

const test        = require("tape")
    , getTokenize = require("../../get-tokenize");

test("Main", t => {
	t.equal(typeof getTokenize, "function", "Supported");
	if (!getTokenize) {
		t.end();
		return;
	}

	t.test("Memoize", t => {
		t.equal(getTokenize("UTC"), getTokenize("UTC"));
		t.end();
	});

	t.test("UTC", t => {
		const tokenize = getTokenize("UTC");

		t.deepEqual(tokenize(new Date(Date.UTC(2017, 1))), {
			year: 2017,
			month: 1,
			date: 1,
			hours: 0,
			minutes: 0,
			seconds: 0,
			milliseconds: 0
		});
		t.end();
	});

	t.test("Europe/Warsaw", t => {
		const tokenize = getTokenize("Europe/Warsaw");

		t.deepEqual(tokenize(new Date(Date.UTC(2017, 1, 2, 17))), {
			year: 2017,
			month: 1,
			date: 2,
			hours: 18,
			minutes: 0,
			seconds: 0,
			milliseconds: 0
		});
		t.deepEqual(
			tokenize(new Date(Date.UTC(2017, 7, 2, 16))),
			{ year: 2017, month: 7, date: 2, hours: 18, minutes: 0, seconds: 0, milliseconds: 0 },
			"DST"
		);
		t.end();
	});

	t.test("Asia/Shanghai", t => {
		const tokenize = getTokenize("Asia/Shanghai");

		t.deepEqual(tokenize(new Date(Date.UTC(2017, 1, 2, 10))), {
			year: 2017,
			month: 1,
			date: 2,
			hours: 18,
			minutes: 0,
			seconds: 0,
			milliseconds: 0
		});
		t.end();
	});

	t.test("Milliseconds", t => {
		const tokenize = getTokenize("Europe/Warsaw");

		t.deepEqual(tokenize(new Date(Date.UTC(2017, 1, 2, 17, 14, 20, 123))), {
			year: 2017,
			month: 1,
			date: 2,
			hours: 18,
			minutes: 14,
			seconds: 20,
			milliseconds: 123
		});
		t.end();
	});

	t.test("Date object", t => {
		const tokenize = getTokenize("Europe/Warsaw")
		    , { dateObject } = tokenize(new Date(Date.UTC(2017, 1, 2, 17, 14, 20, 123)));

		t.ok(dateObject instanceof Date);
		t.deepEqual(
			{
				year: dateObject.getFullYear(),
				month: dateObject.getMonth(),
				date: dateObject.getDate(),
				hours: dateObject.getHours(),
				minutes: dateObject.getMinutes(),
				seconds: dateObject.getSeconds(),
				milliseconds: dateObject.getMilliseconds()
			},
			{
				year: 2017,
				month: 1,
				date: 2,
				hours: 18,
				minutes: 14,
				seconds: 20,
				milliseconds: 123
			}
		);
		t.end();
	});

	t.test("Invalid usage", t => {
		t.throws(() => { getTokenize("WRONG TIMEZONE"); }, RangeError);

		const tokenize = getTokenize("Asia/Shanghai");

		t.throws(() => {
			tokenize(new Date(2014999999999, 99));
		}, /Invalid Date is not valid Date object/u);
		t.throws(() => {
			tokenize(new Date(-10, 1));
		}, /Invalid arguments: No support for BC years/u);
		t.end();
	});
	t.end();
});


index.js
"use strict";

const copyDate     = require("es5-ext/date/#/copy")
    , isValue      = require("es5-ext/object/is-value")
    , ensureNumber = require("es5-ext/object/ensure-finite-number")
    , getTokenize  = require("./get-tokenize");

if (!getTokenize) {
	module.exports = null;
	return;
}

const resolveDate = function (refDate, resultDate, tokenize) {
	// refDate is a local time zone date that expresses the time we're after in the target timezone
	// resultDate is a potentially correct result
	const resolvedDate = tokenize(resultDate).dateObject;
	// resolvedDate shows in local time zone time that resultDate represents in the target timezone
	// If they are the same, we are done
	if (resolvedDate.getTime() === refDate.getTime()) return resultDate;
	resultDate.setTime(resultDate.getTime() + refDate.getTime() - resolvedDate.getTime());
	return resolveDate(refDate, resultDate, tokenize);
};

module.exports = function (timezone) {
	const tokenize = getTokenize(timezone);

	return function (
		yearOrTimeValue,
		month = null,
		date = 1,
		hours = 0,
		minutes = 0,
		seconds = 0,
		milliseconds = 0
	) {
		yearOrTimeValue = ensureNumber(yearOrTimeValue);
		let sampleDate;
		if (isValue(month)) {
			month = ensureNumber(month);
			sampleDate = new Date(
				yearOrTimeValue, month, date, hours, minutes, seconds, milliseconds
			);
		} else {
			sampleDate = new Date(yearOrTimeValue);
		}
		return resolveDate(sampleDate, copyDate.call(sampleDate), tokenize);
	};
};


get-tokenize.js
"use strict";

const ensureDate   = require("es5-ext/date/valid-date")
    , assign       = require("es5-ext/object/assign")
    , ensureString = require("es5-ext/object/validate-stringifiable-value")
    , d            = require("d")
    , lazy         = require("d/lazy");

const refLocale = "en", dateStrRe = /^(\d{2})\/(\d{2})\/(\d{1,4}), (\d{2}):(\d{2}):(\d{2})$/u;

const formatOptions = {
	hour12: false,
	year: "numeric",
	month: "2-digit",
	day: "2-digit",
	hour: "2-digit",
	minute: "2-digit",
	second: "2-digit"
};

// Detect support
if (
	typeof Intl !== "object" ||
	!Intl ||
	typeof Intl.DateTimeFormat !== "function" ||
	typeof Intl.DateTimeFormat.supportedLocalesOf !== "function" ||
	Intl.DateTimeFormat.supportedLocalesOf(refLocale)[0] !== refLocale ||
	new Intl.DateTimeFormat(refLocale, assign({ timeZone: "Europe/Warsaw" }, formatOptions)).format(
		new Date(1496316716561)
	) !== "06/01/2017, 13:31:56" ||
	new Intl.DateTimeFormat(refLocale, assign({ timeZone: "Asia/Shanghai" }, formatOptions)).format(
		new Date(1496316716561)
	) !== "06/01/2017, 19:31:56"
) {
	module.exports = null;
	return;
}

const getFormatter = function (timezone) {
	return new Intl.DateTimeFormat(refLocale, assign({ timeZone: timezone }, formatOptions));
};

const tokenizers = Object.create(null), yearZero = new Date(-1, 12, 2);

const Tokens = function (date, formatter) {
	const match = formatter.format(date).match(dateStrRe);
	this.year = Number(match[3]);
	this.month = Number(match[1]) - 1;
	this.date = Number(match[2]);
	this.hours = Number(match[4]);
	this.minutes = Number(match[5]);
	this.seconds = Number(match[6]);
	this.milliseconds = date.getMilliseconds();
	if (this.hours === 24) this.hours = 0;
};
Object.defineProperties(
	Tokens.prototype,
	lazy({
		dateObject: d(function () {
			return new Date(
				this.year, this.month, this.date, this.hours, this.minutes, this.seconds,
				this.milliseconds
			);
		})
	})
);

module.exports = function (timezone) {
	timezone = ensureString(timezone);
	if (tokenizers[timezone]) return tokenizers[timezone];

	const formatter = getFormatter(timezone);

	return (tokenizers[timezone] = function (date) {
		ensureDate(date);
		if (date < yearZero) {
			throw new TypeError("Invalid arguments: No support for BC years");
		}
		return new Tokens(date, formatter);
	});
};
